Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp

```diff
@@ -635,29 +641,117 @@ RegisterObject::operator int()
                 reg.defaultTemplate = entry.defaultTemplate;
             }
         }
-        for (auto & creator_entry : entry.creatorMap)
+        for (auto& creator_entry : entry.creatorMap)
         {
-            const std::string & template_name = creator_entry.first;
+            const std::string& template_name = creator_entry.first;
             if (reg.creatorMap.find(template_name) != reg.creatorMap.end()) {
                 if (template_name.empty()) {
                     msg_warning("ObjectFactory") << "Class already registered: " << entry.className;
-                } else {
+                }
+                else {
                     msg_warning("ObjectFactory") << "Class already registered: " << entry.className << "<" << template_name << ">";
                 }
-            } else {
+            }
+            else {
                 reg.creatorMap.insert(creator_entry);
             }
         }
 
-        for (const auto & alias : entry.aliases)
+        for (const auto& alias : entry.aliases)
         {
             if (reg.aliases.find(alias) == reg.aliases.end())
             {
-                ObjectFactory::getInstance()->addAlias(alias,entry.className);
+                objectFactory->addAlias(alias, entry.className);
             }
         }
-        return 1;
+        return true;
+    }
+}
+
+typedef struct ObjectRegistrationEntry
+{
+    inline static const char* symbol = "registerObjects";
+    typedef void (*FuncPtr) (sofa::core::ObjectFactory*);
+    FuncPtr func;
+    void operator()(sofa::core::ObjectFactory* data)
+    {
+        if (func) return func(data);
+    }
+    ObjectRegistrationEntry() :func(nullptr) {}
+} ObjectRegistrationEntry;
+
+bool ObjectFactory::registerObjectsFromPlugin(const std::string& pluginName)
```

Comment:
Instead, I suggest the following 
```suggestion
bool ObjectFactory::registerObjectsFromPlugin(const helper::system::Plugin& plugin)
```

In `Main.cpp` , you would need to do:

```cpp
    for (const auto& [_, plugin] : pluginManager.getPluginMap())
    {
        objectFactory->registerObjectsFromPlugin(plugin);
    }
```

Reply:
Actually it was what it was done at the beginning https://github.com/sofa-framework/sofa/blob/c36c1458c968d59275eec5e69b219a9bc8177de8/Sofa/framework/Core/src/sofa/core/ObjectFactory.h#L198

But how do use it in [Sofa/Component/src/sofa/component/init.cpp](https://github.com/sofa-framework/sofa/pull/4429/files/6eacd8030a1f4e532b0ce705464616fc71ef971b#diff-1dd06bf92fc1ac386c2d549a88e5072dbbaa36714904ac62d448f36b531ffebd) ? You would still need to:
 - get the pluginmanager (of course)
 - find the plugin in the pluginMap (with its string ID)
 - then calling registerObjectsFromPlugin on the retrieved pointer (after a nullptr check)

Reply:
I understand. But this is what bothers me. I don't like that a meta-module needs the `PluginManager` (and even a `Plugin` object) to be able to interact with its own dependencies. If we use the `PluginManager` in `Sofa.Component` to load all the sub-modules, it's not even necessary to have them as a dependency of the module.